packages/skott/src/modules/resolvers/ecmascript/resolver.ts

```diff
@@ -112,6 +115,16 @@ export function isJavaScriptModule(module: string): boolean {
   );
 }
 
+export function isTypeScriptProject(tsConfigPath: string) {
+  return pipe(
+      Effect.service(FileReaderTag),
+      Effect.flatMap(
+          (fileReader) => Effect.tryPromise(() => fileReader.stats(tsConfigPath))
```

Comment:
fs.stat throws an error when the file is not found. `Effect.tryPromise` does not swallow the error, instead it produces an untyped error `Effect<FileReader, unknown, boolean>` when there is no file (_unknown_ means *an untyped error can be produced by running this Effect*, it does not mean that no expected errors can occur).

Generally speaking `Effect.tryPromise` is strictly equivalent to 

```ts
Effect.tryCatchPromise(
  () => fileReader.stats(path),
  () => new FileNotFound()
)
```

the main difference being that the latter Effect will have a typed error instead `Effect<FileReader, FileNotFound, boolean>`. But is means that if you run that Effect using `runPromise`, it will throw a Promise Rejection if ever the FileNotFound error is produced.

If we want to return a boolean in all cases, we must use a `catchAll` and recover from the error:

```ts
export function isTypeScriptProject(tsConfigPath: string) {
  return pipe(
      Effect.service(FileReaderTag),
      Effect.flatMap(
          (fileReader) => Effect.tryPromise(() => fileReader.stats(tsConfigPath))
      ),
      Effect.map((fileSize)=> fileSize > 0),
      Effect.catchAll(() => Effect.succeed(false))
  );
}
```

That way the error is correctly handled and we return _false_ when the file is not being found. Also, types can confirm us that the **expected error** is now fully handled -> `Effect<FileReader, never, boolean>` (notice the "never" in the error channel)

Reply:
> fs.stat throws an error when the file is not found

Just to be sure,  the fileReader.stats always returns a number right ? So there is no other error to catch

```typescript
 stats(filename) {
      return fs
          .stat(filename, { bigint: false })
          .then((stats) => stats.size)
          .catch(() => 0);
}
```

Typescript seems to correctly infer the boolean here
<img width="436" alt="image" src="https://github.com/antoine-coulon/skott/assets/9294168/039c19bb-93a6-47aa-b37c-b4f405f6f6f0">

Reply:
> Just to be sure, the fileReader.stats always returns a number right ? So there is no other error to catch

Oh yeah you're right my bad, I didn't remember that `fs.stat` implementation was safe towards errors, so it's fine :) 

> Typescript seems to correctly infer the boolean here

This is the danger of `runPromise` actually, as Promises only have one generic the type information of error is swallowed so you lose all type of information apart the success channel (last Effect generic) which is inferred in the Promise generic, but it does not mean that the error is not there anymore.

```ts
const failingEffect = Effect.fail("error") // Effect<never, string, void>;

const result = Effect.runPromise(failingEffect); // Will throw an uncaught promise rejection, even if it's a `Promise<void>`
```

That's also why `runPromiseExit` exists, it basically wraps the result into an Exit (Either-like) instead of unwrapping in a unsafe way:

```ts
const failingEffect = Effect.fail("error") // Effect<never, string, void>;

const exit = Effect.runPromiseExit(failingEffect); // Will not throw an return a Promise<Exit<E, A>>;
```

That way you can then pattern match on either the success or the failure:

```ts
const result = await exit;

pipe(
  result,
  Exit.match(
    (error) => // doSomething with it
    (result) => // in that case void
  )
)
```